thirdeye/thirdeye-frontend/app/pods/services/rootcause-fetcher/service.js

```diff
@@ -0,0 +1,210 @@
+      this._swap(node, parent(node));
+      node = parent(node);
+    }
+  }
+
+  _siftDown() {
+    let node = top;
+    while (
+      (left(node) < this.size() && this._greater(left(node), node)) ||
+      (right(node) < this.size() && this._greater(right(node), node))) {
+      let maxChild = (right(node) < this.size() && this._greater(right(node), left(node))) ? right(node) : left(node);
+      this._swap(node, maxChild);
+      node = maxChild;
+    }
+  }
+}
+
+/**
+ * Promise-lookalike for manual scheduling and execution via service
+ */
+class MyPromise {
+  constructor (url, priority, externalOrder, internalOrder) {
+    this.url = url;
+    this.priority = priority;
+    this.externalOrder = externalOrder;
+    this.internalOrder = internalOrder;
+
+    this.result = null;
+    this.error = null;
+    this.reject = null;
+    this.resolve = null;
+
+    this.delegate = new RSVP.Promise((resolve, reject) => {
+      this.reject = reject;
+      this.resolve = resolve;
+    });
+  }
+
+  then(args) {
+    return this.delegate.then(args);
+  }
+
+  catch(args) {
+    return this.delegate.catch(args);
+  }
+}
+
+/**
+ * Query scheduler based on priority, external, and internal order. Supports configurable
+ * make span.
+ */
+export default Service.extend({
+  parallelism: 12,
+
+  counter: 0,
+
+  waiting: null,
+
+  running: null,
+
+  init() {
+    this._super(...arguments);
+    setProperties(this, { counter: 0, waiting: new PriorityQueue(ROOTCAUSE_FETCHER_COMPARATOR), running: [] });
+  },
+
+  fetch(url, priority = 0, externalOrder = 0) {
+    const counter = get(this, 'counter');
+
+    const promise = new MyPromise(url, priority, externalOrder, counter);
+
+    set(this, 'counter', counter + 1);
+    get(this, 'waiting').push(promise);
+
+    this._update();
+
+    return promise.delegate;
+  },
+
+  reset() {
+    set(this, 'waiting', new PriorityQueue(ROOTCAUSE_FETCHER_COMPARATOR));
+  },
+
+  resetPrefix(urlPrefix) {
+    const waiting = get(this, 'waiting');
+
+    const newWaiting = new PriorityQueue(ROOTCAUSE_FETCHER_COMPARATOR);
+    while (!waiting.isEmpty()) {
+      if (!waiting.peek().url.startsWith(urlPrefix)) {
+        newWaiting.push(waiting.pop());
+      }
+    }
+
+    set(this, 'waiting', newWaiting);
+  },
+
+  _update() {
+    const { waiting, running, parallelism } =
+      getProperties(this, 'waiting', 'running', 'parallelism');
+
+    const remaining = running.filter(p => p.result === null && p.error === null);
+
+    const nStarting = parallelism - remaining.length;
+    const starting = [];
+    for (let i = 0; i < nStarting && !waiting.isEmpty(); i++) {
+      starting.pushObject(waiting.pop()); // modified in-place
+    }
+
+    setProperties(this, { running: [...remaining, ...starting] });
+
+    starting.forEach(p => {
+      fetch(p.url)
+        .then(res => { p.result = res; this._update(); })
+        .catch(err => { p.error = err; this._update(); });
+    });
```

Comment:
I would swap this forEach for `for..of` and add async/await if you want guarantee for the order of return responses and also if there is dependencies of something here (ex. p.result) in the codes after it. 

`
  async _update() {
..
 for (const p of starting) {
      await fetch(p.url)
        .then(res => { p.result = res; this._update(); })
        .catch(err => { p.error = err; this._update(); });
    });`

Reply:
This is all about concurrency